sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java

```diff
@@ -198,50 +498,20 @@ public static Object convertAvroFieldStrict(
         throw new IllegalArgumentException("Can't convert 'null' to non-nullable field");
 
       default:
-        throw new AssertionError("Unexpected AVRO Schema.Type: " + unwrapped.getType());
+        throw new AssertionError("Unexpected AVRO Schema.Type: " + type.type.getType());
     }
   }
 
-  @VisibleForTesting
-  static org.apache.avro.Schema unwrapNullableSchema(org.apache.avro.Schema avroSchema) {
-    if (avroSchema.getType() == org.apache.avro.Schema.Type.UNION) {
-      List<org.apache.avro.Schema> types = avroSchema.getTypes();
-
-      // optional fields in AVRO have form of:
-      // {"name": "foo", "type": ["null", "something"]}
-
-      // don't need recursion because nested unions aren't supported in AVRO
-      List<org.apache.avro.Schema> nonNullTypes =
-          types
-              .stream()
-              .filter(x -> x.getType() != org.apache.avro.Schema.Type.NULL)
-              .collect(Collectors.toList());
-
-      if (nonNullTypes.size() == types.size()) {
-        // union without `null`, keep as is
-        return avroSchema;
-      } else if (nonNullTypes.size() > 1) {
-        return org.apache.avro.Schema.createUnion(nonNullTypes);
-      } else if (nonNullTypes.size() == 1) {
-        return nonNullTypes.get(0);
-      } else { // nonNullTypes.size() == 0
-        return avroSchema;
-      }
-    }
-
-    return avroSchema;
-  }
-
   private static Object convertRecordStrict(GenericRecord record, Schema.FieldType fieldType) {
     checkTypeName(fieldType.getTypeName(), Schema.TypeName.ROW, "record");
-    return toRowStrict(record, fieldType.getRowSchema());
+    return toBeamRowStrict(record, fieldType.getRowSchema());
   }
 
   private static Object convertBytesStrict(ByteBuffer bb, Schema.FieldType fieldType) {
     checkTypeName(fieldType.getTypeName(), Schema.TypeName.BYTES, "bytes");
 
     byte[] bytes = new byte[bb.remaining()];
-    bb.get(bytes);
+    bb.duplicate().get(bytes);
```

Comment:
Good catch! As I understand, `.get(byte[])` will mutate ByteBuffer `position` and it will fail on the second attempt.

Reply:
I see the same issue in `BigQueryAvroUtils`

Reply:
Probably not causing issues in BigQueryAvroUtils as I don't think those Avro objects are ever reused. However maybe we should remove the code in BigQueryAvroUtils and just standardize on this library.